src/query/builder.js

```diff
@@ -68,6 +68,44 @@ assign(Builder.prototype, {
     return this;
   },
 
+  // With
+  // ------
+
+  with(alias, statement, bindings) {
+    if (typeof statement === 'function') {
+      return this.withWrapped(alias, statement);
+    }
+    // Allow a raw statement to be passed along to the query.
+    if (statement instanceof Raw && arguments.length >= 2) {
+      return this.withRaw(alias, statement, bindings);
+    }
+
```

Comment:
Yep, could be even more detailed describing that argument must be either function / raw. There are no exceptions thrown in this file currently, but actually many of the function does throw an error if called with invalid parameters.